src/axom/primal/geometry/BezierCurve.hpp

```diff
@@ -0,0 +1,335 @@
+// Copyright (c) 2017-2019, Lawrence Livermore National Security, LLC and
+// other Axom Project Developers. See the top-level COPYRIGHT file for details.
+//
+// SPDX-License-Identifier: (BSD-3-Clause)
+
+/*!
+ * \file BezierCurve.hpp
+ *
+ * \brief A BezierCurve primitive
+ */
+
+#ifndef PRIMAL_BEZIERCURVE_HPP_
+#define PRIMAL_BEZIERCURVE_HPP_
+
+#include "axom/core.hpp"
+#include "axom/slic.hpp"
+
+#include "axom/primal/geometry/NumericArray.hpp"
+#include "axom/primal/geometry/Point.hpp"
+#include "axom/primal/geometry/Vector.hpp"
+#include "axom/primal/geometry/Segment.hpp"
+#include "axom/primal/geometry/BoundingBox.hpp"
+#include "axom/primal/geometry/OrientedBoundingBox.hpp"
+
+#include "axom/primal/operators/squared_distance.hpp"
+
+#include "fmt/fmt.hpp"
+#include <vector>
+#include <ostream>
+
+namespace axom
+{
+namespace primal
+{
+
+// Forward declare the templated classes and operator functions
+template < typename T, int NDIMS >
+class BezierCurve;
+
+/*! \brief Overloaded output operator for Bezier Curves*/
+template < typename T,int NDIMS >
+std::ostream& operator<<(std::ostream & os,
+                         const BezierCurve< T,NDIMS > & bCurve);
+
+/*!
+ * \class BezierCurve
+ *
+ * \brief Represents a Bezier curve defined by an array of control points
+ * \tparam T the coordinate type, e.g., double, float, etc.
+ * \tparam NDIMS the number of dimensions
+ *
+ * The order of a Bezier curve with N+1 control points is N.
+ * The curve is approximated by the control points,
+ * parametrized from t=0 to t=1.
+ */
+
+template < typename T,int NDIMS >
+class BezierCurve
+{
+public:
+  using PointType = Point< T,NDIMS >;
+  using VectorType = Vector< T,NDIMS >;
+  using NumArrayType = NumericArray< T,NDIMS >;
+  using SegmentType = Segment< T, NDIMS >;
+  using CoordsVec = std::vector< PointType >;
+  using BoundingBoxType = BoundingBox< T, NDIMS >;
+  using OrientedBoundingBoxType = OrientedBoundingBox< T, NDIMS >;
+
+public:
+
+  /*!
+   * \brief Constructor for a Bezier Curve that reserves space for
+   *  the given order of the curve
+   *
+   * \param [in] order the order of the resulting Bezier curve
+   * \pre order is greater than or equal to -1.
+   */
+  explicit BezierCurve(int ord = -1)
```

Comment:
Could the default be set to zero and simplify the code below?

Reply:
I don't think this is a good idea.

An order-zero Bezier curve has a single control point, i.e. it is a point.
If the user does not provide this point, then it will be filled with garbage data.

Reply:
Do you think it's a good idea to create a Bezier curve that is order less than zero?

Reply:
Yes, if the user doesn't have data to fill it. 

It would be arbitrary and error prone to set this to zero, which would imply that there is a valid control point.

Reply:
OK, I see, so this serves as a sentinel that the Bezier curve is not properly initialized...